enhancements/user-workload-monitoring.md

```diff
@@ -0,0 +1,275 @@
+Currently it is difficult to achieve metering/chargeback for Red Hat products without a central observability stack.
+
+### Goals
+
+Allow deployment and reconciliation of scrape targets, recording rules, and alerting rules in user namespaces.
+This effectively means the following custom resources:
+
+-  [ServiceMonitor](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#servicemonitor)
+- [PodMonitor](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#podmonitor)
+- [PrometheusRule](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#prometheusrule)
+
+Allow the correlation between app and cluster centric metrics.
+
+### Non-Goals
+
+Out of scope are:
+
+- Definition of Dashboards.
+- Deployment of Prometheus instances in user namespaces.
+- Deployment of Alertmanager instances in user namespaces.
+
+## Proposal
+
+![](user-monitoring-deployment.png)
+
+The existing cluster monitoring stack is going to be extended with additional Prometheus (P) servers
+and one additional Prometheus Operator (PO) reconciling user namespaces
+as well as the newly added Prometheus servers.
+
+The cluster Prometheus Operator (PO) continues to operate the cluster-level Prometheus.
+We will continue to use a centralized Alertmanager (A) cluster.
+The existing Alertmanager cluster will aggregate both user workload alerts as well as cluster alerts. Tenancy will be achieved by forcing specific labels onto metrics and alerts by which the central Alertmanager config can route similarly to how tenancy is achieved at the query level (see below).
+
+Querying data from a single, multi-tenant interface is done using the Thanos Querier (TQ) component.
+Tenancy is enforced at the prometheus query layer. This is achived by leveraging the existing topology using kube-rbac-proxy [1] and prom-label-proxy [2].
+
+Differences between cluster- and user-level is mainly upgrade guarantees. The cluster-level Prometheus stack is an integral part of OpenShift and its reliability is important, keeping the stack immutable and fully testable by us allows us to rely on it to a higher degree. As we have no such possibility for user defined configurations and scrape targets, we can't provide the same guarantees.
+
+There are edge cases where an automated upgrade is not possible in the face of user workload monitoring.
+If users declare i.e. queries in ServiceMonitors which break in newer Prometheus versions
+we most likely cannot perform an automated migration. Such migration without knowledge of user intention behind each query would be exceptionally difficult.
+
+Instead, the "upgradable" condition in the ClusterOperator will be set to false.
+Routing configurations for alerts to be sent to specific teams need to be communicated/configured by an infrastructure owner using the existing Alertmanager stack that is part of cluster-monitoring.
+We will look into extension mechanisms post GA.
```

Comment:
Can I push back a bit here and say that extension mechanisms, particularly for the alertmanager approach, be part of the GA criteria?

Reply:
That's good feedback. We'll communicate this with product.

Reply:
cc @sichvoge

Reply:
@jeremyeder what is your expectation for this?

Reply:
That there is no requirement for 'cluster-admin' permissions in order to use the user-workload monitoring stack.

Reply:
@jeremyeder that is a very good suggestion :+1:

Reply:
I understand why you want that and I can see it be very useful for OSD, but we have so many items on our GA list that we can't promise it to be in. My suggestion would be to have it as a stretch goal and only implement it if all the other items are done.

TD;LR: Not in the acceptance criteria for 4.4, but we can add it as a stretch goal.